k2/csrc/array.h

```diff
@@ -1,69 +1,113 @@
 // k2/csrc/array.h
 
 // Copyright (c)  2020  Xiaomi Corporation (author: Daniel Povey)
+//                      Fangjun Kuang (csukuangfj@gmail.com)
 
 // See ../../LICENSE for clarification regarding multiple authors
 
 #ifndef K2_CSRC_ARRAY_H_
 #define K2_CSRC_ARRAY_H_
 
-#include <functional>
-#include <limits>
-#include <memory>
+#include <type_traits>
 #include <vector>
 
+#include "glog/logging.h"
+
 namespace k2 {
 
 /*
-   We will use e.g. StridedPtr<int32_t, T> when the stride is not 1, and
+   We will use e.g. StridedPtr<T*, int32_t> when the stride is not 1, and
    otherwise just T* (which presumably be faster).
 */
-template <typename I, typename T>
+template <typename Ptr, typename I = int32_t>
 struct StridedPtr {
-  T *data;
-  I stride;
-  T &operator[](I i) { return data[i]; }
+  using T = typename std::remove_pointer<Ptr>::type;
+
+  T *data;   // it is NOT owned here
+  I stride;  // in number of elements, NOT number of bytes.
+
+  T &operator[](I i) { return data[i * stride]; }
+  const T &operator[](I i) const { return data[i * stride]; }
+
   StridedPtr(T *data, I stride) : data(data), stride(stride) {}
```

Comment:
need  `operator ==`  to compare with `null_ptr` as you called it in `Array1/2`

Reply:
also a constructor `StridedPtr(T* data)` as we call `data_(ptr)` in `Array1`

Reply:
A conversion operator is added for `StridedPtr`.